torch/csrc/api/include/torch/nn/modules/rnn.h

```diff
@@ -1,121 +1,113 @@
 #pragma once
 
-#include <torch/detail.h>
 #include <torch/nn/module.h>
 
 #include <torch/csrc/autograd/variable.h>
 
 #include <ATen/Error.h>
+#include <ATen/optional.h>
 
 #include <cstdint>
+#include <functional>
 #include <memory>
 #include <vector>
 
+namespace torch { namespace nn {
+class Dropout;
+}} // namespace torch::nn
+
 namespace torch { namespace nn {
 template <typename Derived>
 class RNNBase : public CloneableModule<Derived> {
  public:
-  // These must line up with the CUDNN mode codes
-  enum RNNMode { RNN_RELU = 0, RNN_TANH = 1, LSTM = 2, GRU = 3 };
+  // These must line up with the CUDNN mode codes:
+  // https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnRNNMode_t
+  enum class CuDNNMode { RNN_RELU, RNN_TANH, LSTM, GRU };
 
   RNNBase(
-      uint32_t input_size,
-      uint32_t hidden_size,
-      int mode,
-      uint32_t nlayers,
-      bool with_bias,
-      float dropout);
-
-  using CloneableModule<Derived>::parameters;
-  using CloneableModule<Derived>::is_training;
+      int64_t input_size,
+      int64_t hidden_size,
+      at::optional<CuDNNMode> cudnn_mode = at::nullopt,
+      size_t number_of_gates = 1,
+      bool has_cell_state = false);
 
-  bool flatten_parameters(); // Flatten for cudnn
+  void reset() override;
 
   variable_list forward(variable_list) override;
 
-  void cpu() override;
-  void cuda() override;
+  void to(at::Type& type) override;
+  void to(at::ScalarType scalar_type) override;
+  void to(at::Backend backend) override;
 
-  std::vector<Variable> ihw;
-  std::vector<Variable> ihb;
-  std::vector<Variable> hhw;
-  std::vector<Variable> hhb;
+  TORCH_PARAMETER(int64_t, input_size);
+  TORCH_PARAMETER(int64_t, hidden_size);
+  TORCH_PARAMETER(int64_t, layers) = 1;
+  TORCH_PARAMETER(bool, with_bias) = true;
+  TORCH_PARAMETER(double, dropout) = 0.0;
 
- private:
-  using CloneableModule<Derived>::add;
-
-  uint32_t input_size_;
-  uint32_t hidden_size_;
-  uint32_t gate_size_;
-  RNNMode mode_;
-  uint32_t nlayers_;
-  bool with_bias_;
-  float dropout_;
-
-  // This is copied from pytorch, to determine whether weights are flat for
-  // the fast CUDNN route. Otherwise, we have to use non flattened weights,
-  // which
+ protected:
+  virtual variable_list cell_forward(variable_list, size_t layer) = 0;
+
+  variable_list CUDNN_forward(variable_list);
+  variable_list autograd_forward(variable_list);
+
+  void flatten_parameters_for_cudnn();
```

Comment:
Shouldn't this be public? How will I call it if I mess with the weights of the LSTM?

Reply:
The answer I'd like to give here is that if you look at the cursor implementation from https://github.com/pytorch/pytorch/pull/6345, you'll be able to write `module.parameters().find("weights")` and you'll get back the parameter if it's there

Reply:
I understand, but I can't flatten the parameters if I do something like point the internal weight tensor else where, and then my LSTM will always take the slowering non-cudnn route.

Reply:
Oh sorry, I answered a different question :P I will make this method public again